relay/pkg/block/pull.go

```diff
@@ -0,0 +1,77 @@
+package block
+
+import (
+	"fmt"
+
+	"github.com/keep-network/tbtc/relay/pkg/btc"
+)
+
+func (f *Forwarder) pullNextHeader() error {
+	nextHeader, err := f.pullHeaderFromBtcChain(f.nextPullHeaderHeight)
+	if err != nil {
+		return fmt.Errorf(
+			"could not get header by height at [%d]: [%v]",
+			f.nextPullHeaderHeight,
+			err,
+		)
+	}
+
+	// TODO: Consider just comparing hashes - should be enough
+	// and check if nextHeader and lastPulledHeader header can ever be not equal
+	if !nextHeader.Equals(f.lastPulledHeader) {
+		f.pushHeaderToQueue(nextHeader)
+		f.lastPulledHeader = nextHeader
+		f.nextPullHeaderHeight++
+	}
+
+	return nil
+}
+
+func (f *Forwarder) pullHeaderFromBtcChain(height int64) (*btc.Header, error) {
+	return f.btcChain.GetHeaderByHeight(height)
+}
+
+func (f *Forwarder) pushHeaderToQueue(header *btc.Header) {
+	f.headersQueue <- header
+}
```

Comment:
Regarding those guys, the problem is as follows:
- `pullNextHeader` does too much. It's name suggests it only pulls next header but it actually pushes it to the queue and sets some internal forwarder state.
- `pullHeaderFromBtcChain` and `pushHeaderToQueue` don't do anything deserving a separate method

I'd rather go that way:
```
func (f *Forwarder) pullHeaderFromBtcChain(
	ctx context.Context,
) (*btc.Header, error) {
	for {
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}

		chainHeight, err := f.btcChain.GetBlockCount()
		if err != nil {
			return nil, fmt.Errorf("could not get block count [%v]", err)
		}

		// Check if there are more headers to pull or we are above the chain's
		// tip and need to sleep until the chain adds more headers
		if f.nextPullHeaderHeight <= chainHeight {
			nextHeader, err := f.btcChain.GetHeaderByHeight(f.nextPullHeaderHeight)
			if err != nil {
				return nil, fmt.Errorf(
					"could not get header by height at [%d]: [%v]",
					f.nextPullHeaderHeight,
					err,
				)
			}

			// TODO: Consider just comparing hashes - should be enough
			//  and check if nextHeader and lastPulledHeader header can ever
			//  be not equal
			if !nextHeader.Equals(f.lastPulledHeader) {
				return nextHeader, nil
			}
		}

		select {
		case <-time.After(forwarderPullingSleepTime):
		case <-ctx.Done():
		}
	}
}

func (f *Forwarder) pushHeaderToQueue(header *btc.Header) {
	f.headersQueue <- header
	f.lastPulledHeader = header
	f.nextPullHeaderHeight++
}
```
That way we solve all the problems and we obtain a clean pulling loop. All pull logic lives in `pull.go` and reading from BTC chain is separated from pushing to the queue. What do you think?

Reply:
I'm ok with that.